[repository-s3] Enable S3SignerType

xuzha opened the pull request:
This PR enable user with S3SignerType, which is v2 signer type some of S3 compatible services but without V4 signer type would need this setting.


Related to https://github.com/elastic/elasticsearch/issues/13332

dadoonet commented:
So if I understand correctly, S3Signer is specific to S3 endpoints.
Does it mean that AWSxSignerType are specific to EC2 endpoints?

Any chance you can fill this table with `X` where a type is available for a given API?

| Signer  | EC2  | S3  |
|:----------:|:------:|:-----:|
| QueryStringSignerType  |   |   |
| AWS3SignerType  |   |   |
| AWS4SignerType  |   |   |
| NoOpSignerType  |   |   |
| S3SignerType  |   |  X |

Thanks!

xuzha commented:
@dadoonet  sure. Yes from the aws-sdk, S3SignerType is only for S3, [source](https://github.com/aws/aws-sdk-java/blob/1.10.12/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/AmazonS3Client.java#L275) . This is indeed a very hacky solution and Beijing and Frankfurt region don't support it. But there's not much we can do about it for these service only support V2. Do you think we should log a warning msg here?

And the form : 

| Signer  | EC2  | S3  |
|:----------:|:------:|:-----:|
| QueryStringSignerType  | ?  | ? |  
| AWS3SignerType  |X | X |
| AWS4SignerType  | X  | X  |
| NoOpSignerType  | X  | X  |
| S3SignerType  |   |  X |
| AWSS3V4SignerType  |   |  X |

For QueryStringSignerType, both EC2 and S3 client accept them. I think QueryStringSignerType should be used only for AWS query string request. I think we could it there let user decided if they want to use it.

dadoonet commented on plugins/repository-s3/src/main/java/org/elasticsearch/cloud/aws/AwsSigner.java:
```diff
@@ -34,20 +34,29 @@ private AwsSigner() {
+            // V2 signer type is not support in aws China and Frankfurt region
+            if (endpoint.equals("s3.cn-north-1.amazonaws.com.cn") ||
+                    endpoint.equals("s3.eu-central-1.amazonaws.com")) {
+                throw new IllegalArgumentException(endpoint + " does not support S3SignerType");
```
I think we should only log.warn here. I can imagine that at some point AWS may support this in China and I would not block users for this.
May be we should not control that at all and let the user specify whatever he wants.

I mean that if this plugin is used with other S3 compatible platform, we can't do those checks.

dadoonet commented on plugins/repository-s3/src/main/java/org/elasticsearch/cloud/aws/AwsSigner.java:
```diff
@@ -34,20 +34,29 @@ private AwsSigner() {
+                // It throws a IllegalArgumentException if not found
+                SignerFactory.getSignerByTypeAndService(signer, null);
+            } catch (IllegalArgumentException e) {
+                throw new IllegalArgumentException("wrong signer set for [cloud.aws.s3.signer] or [cloud.aws.signer]: " + signer);
```
Should we actually control this? I know, I wrote that code as well at the begining but the more I think about it, the more I think it's not that useful to block...

May be only log.warn with "[{}] signer might not be supported"...

xuzha replied:
Hmm, yes, I agree with you, we should not block user with any setting.

dadoonet commented:
I left some comments. 

I think we should also document it. I might have forgotten that part when I split AWS plugin in 2 plugins.

Could you change the following?

* https://github.com/elastic/elasticsearch/blob/master/docs/plugins/repository-s3.asciidoc#L3-2 (add the section which is in ec2 doc here)
* https://github.com/elastic/elasticsearch/blob/master/docs/plugins/discovery-ec2.asciidoc#L113-113 (remove reference to S3 there)

xuzha commented:
thanks for the review @dadoonet 

I updated the PR. Right now it only logging a warning message rather than block user. 
I also updated the signer part in the document.

dadoonet commented on docs/plugins/repository-s3.asciidoc:
```diff
@@ -113,11 +113,14 @@ The available values are:
-with the right signer to use. Defaults to `AWS4SignerType`.
+If you are using a S3 compatible service, they might be using an older API to sign the requests.
+You can set your compatible signer API using `cloud.aws.signer` (or `cloud.aws.s3.signer`) with the right
+signer to use. Defaults to `AWS4SignerType`.
```
I saw in https://github.com/elastic/elasticsearch-cloud-aws/issues/242 that this is wrong actually.
And the code says that default is actually `null` or not specified... I would remove here `Defaults to ` or replace it with `Defaults to not specified`

dadoonet commented on docs/plugins/discovery-ec2.asciidoc:
```diff
@@ -110,10 +110,10 @@ The available values are:
-You can set your compatible signer API using `cloud.aws.signer` (or `cloud.aws.ec2.signer` and `cloud.aws.s3.signer`)
+If you are using a compatible EC2 service, they might be using an older API to sign the requests.
+You can set your compatible signer API using `cloud.aws.signer` (or `cloud.aws.ec2.signer`)
 with the right signer to use. Defaults to `AWS4SignerType`.
```
I saw in https://github.com/elastic/elasticsearch-cloud-aws/issues/242 that this is wrong actually.
And the code says that default is actually `null` or not specified... I would remove here `Defaults to ` or replace it with `Defaults to not specified`

dadoonet commented on docs/plugins/repository-s3.asciidoc:
```diff
@@ -113,11 +113,14 @@ The available values are:
+signer to use. Defaults to `AWS4SignerType`.
+
+If you are using a compatible S3 service which do not support Version 4 signing process, you may need to
+use `S3SignerType`, which is Signature Version 2
```
Just add a dot `.` at the end.

dadoonet commented on plugins/discovery-ec2/src/main/java/org/elasticsearch/cloud/aws/AwsSigner.java:
```diff
@@ -45,9 +41,24 @@ public static void configureSigner(String signer, ClientConfiguration configurat
+
+        try {
+            validateSignerType(signer);
+        }catch(IllegalArgumentException e){
```
`} catch(IllegalArgumentException e) {`

dadoonet commented on plugins/discovery-ec2/src/main/java/org/elasticsearch/cloud/aws/AwsSigner.java:
```diff
@@ -45,9 +41,24 @@ public static void configureSigner(String signer, ClientConfiguration configurat
+     * @param signer Signer to use
+     * @param configuration AWS Client configuration
+     */
+    public static void configureSigner(String signer, ClientConfiguration configuration){
```
`public static void configureSigner(String signer, ClientConfiguration configuration) {`

dadoonet commented on plugins/discovery-ec2/src/test/java/org/elasticsearch/cloud/aws/AWSSignersTests.java:
```diff
@@ -35,6 +35,14 @@ public void testSigners() {
+        assertThat(signerTester("AWSS3V4SignerType"), is(false));
+
+        ClientConfiguration configuration = new ClientConfiguration();
+        AwsSigner.configureSigner("AWS4SignerType", configuration);
```
`assertEquals(configuration.getSignerOverride(), "AWS4SignerType");` ?

dadoonet commented on plugins/repository-s3/src/main/java/org/elasticsearch/cloud/aws/AwsSigner.java:
```diff
@@ -21,33 +21,54 @@
+            default:
+                try {
+                    SignerFactory.getSignerByTypeAndService(signer, null);
+                }catch (IllegalArgumentException e) {
```
`} catch (IllegalArgumentException e) {`

dadoonet commented on plugins/repository-s3/src/main/java/org/elasticsearch/cloud/aws/AwsSigner.java:
```diff
@@ -21,33 +21,54 @@
-        } catch (IllegalArgumentException e) {
-            throw new IllegalArgumentException("wrong signer set [" + signer + "]");
+            validateSignerType(signer, endpoint);
+        }catch(IllegalArgumentException e){
```
`} catch(IllegalArgumentException e) {`

dadoonet commented on plugins/repository-s3/src/test/java/org/elasticsearch/cloud/aws/AWSSignersTests.java:
```diff
@@ -35,8 +35,32 @@ public void testSigners() {
+    public void testV2InInvalidRegion() {
+        try {
+            AwsSigner.validateSignerType("S3SignerType", "s3.cn-north-1.amazonaws.com.cn");
+            fail("S3SignerType should not be able to configured in China region");
```
`S3SignerType should not be available for China region`

dadoonet commented on plugins/repository-s3/src/test/java/org/elasticsearch/cloud/aws/AWSSignersTests.java:
```diff
@@ -35,8 +35,32 @@ public void testSigners() {
+
+        try {
+            AwsSigner.validateSignerType("S3SignerType", "s3.eu-central-1.amazonaws.com");
+            fail("S3SignerType should not be able to configured Frankfurt region");
```
`S3SignerType should not be available for Frankfurt region`

dadoonet commented:
I left some new comments. I think we are super close now.

xuzha commented:
@dadoonet  thanks, updated.

dadoonet commented:
LGTM! Thanks!

dadoonet commented:
Note that if we want to push that in 2.x branch, it will require you some work as we split cloud-aws plugin only in master branch.

xuzha commented:
Thanks  @dadoonet.
For 2.x branch, I did slightly differently here to avoid making the change complicated. https://github.com/elastic/elasticsearch/commit/736e21f021301c2b13f891a056e4f35ebd99dba8.  Let me know if you think we need more changes for 2.x.

dadoonet commented:
That's fine I think. Thanks for fixing that!
